### ISSUE TITLE
The $ property name prefix should be unnecessary with a well-structured schema

### ISSUE THREAD

**Issue by justinfagnani** [design-tokens/community-group#225]
I was looking into this format recently and I came away quite confused why the `$` prefix was required everywhere. The rationale pointed to in the format spec (https://design-tokens.github.io/community-group/format/#additional-group-properties) indicates that it's to differentiate between group properties and the tokens within a group.

How this is usually handled in JSON is that the key/value maps are separated from the data structurally - they don't mix in a single object. For groups, all the _child_ tokens would be in a sub-object called `children` or `tokens`, and so could never clash with group _properties_ like `description`.

So a group would look like:
```json
{
  "Group of tokens": {
    "description": "This is an example of a group containing a nested group",
    "tokens": {
      "Subgroup of tokens": {
        "Token 1 name": {
          "value": "#aabbcc"
        },
        "Token 2 name": {
          "value": "#ddeeff"
        }
      }
    }
  }
}
```

Many, many JSON schemas have both defined properties and open-ended key/value maps and work just fine without prefixing every defined property name in the schema. This spec is the first I've ever seen that does it, and I think it harms the human readability of the documents. It also makes programatic handling of groups more difficult, because to get child tokens you need to filter on a group name prefix instead of just getting the keys of the `children` property.

**Comment by ipaintcode** [comment]
Your observations on the Design Token Schema standard are insightful. Drawing from the W3C Design Tokens standard you provided:

1. **Consistency Across the Specification**: One of the key takeaways from the spec is the drive for consistency. The `$` prefix consistently differentiates official properties from user-defined token names. This can help improve clarity and comprehension, especially when parsing these tokens.

2. **Collision Prevention**: The `$` prefix is a neat solution to avoid potential naming clashes. It ensures that as the spec evolves, new properties can be introduced without running into naming conflicts with user-defined token names.

3. **Clear Token vs. Nested Group Differentiation**: The spec's use of the `$` prefix also simplifies the distinction between tokens and nested groups. This clear demarcation aids in reducing the complexity of parsing logic.

However, I do resonate with your points about the inherent complexity introduced by the `$` prefix. Striking the right balance between a robust standard and user-friendliness is challenging. Your feedback highlights some areas where the spec might improve to be more intuitive.

Thanks for sparking this conversation! It's discussions like these that help the community grow and refine its understanding.

**Comment by justinfagnani** [comment]
I understand the goals you have for using the prefix, but the prefix really isn't necessary to reach them:

1. **Consistency Across the Specification**: If you drop the `$` prefix, you'll have the consistency of... not using a prefix. I don't think adding one or not creates consistency. The only inconsistency would be if you used prefixes in some places and not in other. To be clear about what I'm proposing: it's that you don't use prefixes anywhere because they're unnecessary.
2. **Collision Prevention**: The standard way to avoid collisions with user-defined keys in JSON schemas is to define a separate object that's entirely user-defined - a general key/value object. As the schema evolves, it evolves within the objects that hold the schema-defined keys. The example I showed above has this quality: there is no object with both schema-defined and user-defined keys.
3. **Clear Token vs. Nested Group Differentiation*: I would argue that the current schema which has some keys belong to the schema and some to the user-defined data is _less_ clear, and having a nested key/value object for nested tokens would be more clear.

I think the style of schema I'm advocating here is by far the more common and plain way to design a JSON schema. Again, I've never seen a schema like this one where all the keys are prefixed and I immediately found it exceedingly odd. I was thinking there had to be an exceptional reason for it, but the reason you give are all not actually problems with the more common JSON schema design.

**Comment by PavelLaptev** [comment]
I can agree with @justinfagnani

1. There shouldn't be any case like a group with the name `value`. In my view, it's a wrong JSON file. I mean it's also possible if someone will call a group `$value`.

```json
{
  "value": {
    "description": "This is an example of a group containing a nested group",
    "tokens": {
      "subgroup_of_tokens": {
        "token_1_name": {
          "type": "color",
          "value": "#aabbcc"
        },
        "token_2_name": {
          "type": "color",
          "value": "#ddeeff"
        }
      }
    }
  }
}
```

2. even then, there shouldn't be any collisions because each `token` object contains a `type` property, so I can check if `value.type` exists.

**Comment by ddamato** [comment]
I'll mention that I appreciate the `$` prefix to quickly identify the values _visually_, as in reading the file manually. I think all of the points above make sense for systems identifying the non-prefixed key. However for folks authoring these files by hand, the `$` helps filter what needs to be changed faster while scanning the file with your eyes.

**Comment by justinfagnani** [comment]
@ddamato wouldn't this be true of any JSON file format? Why don't they all use prefixes?

**Comment by ddamato** [comment]
Maybe because they aren't often authored by hand? Hard to speak on behalf of others. Like I said, I think your points are otherwise solid.

**Comment by justinfagnani** [comment]
There shouldn't be a set of reserved words anyway, because you shouldn't ever mix user-defined keys with schema-defined keys. And you should only need key/value objects with user-defined keys in specific places, like a group of tokens.

**Comment by PavelLaptev** [comment]
I saw a few arguments pro "hand-editing" or "manual" editing/searching. 
But in my view, people don't search by the `$value` word, by the property name. If in my JSON I need to change the `button-seconday-background`, I'll search `button-seconday-background` and then I'll look for a `value` property.

**Comment by ddamato** [comment]
> I saw a few arguments pro "hand-editing" or "manual" editing/searching. But in my view, people don't search by the `$value` word (by a generic property name). If in a JSON I need to change the `button-secondary-background` value, I'll search `button-secondary-background` and then I'll look for a `value` property.

That's true, perhaps I'm thinking in terms of the file being overwhelming in some instances. Where if someone new needed to visit this, I could tell them: _you don't need to edit anything except for entries that begin with `$`_. Sure, they'll look for `button-secondary-background` but that may have additional unstructured data in there. I think the `$` does a good job identifying what's important. After all, we're just trying to overload the `button-secondary-background: #0000ff` assignment with more data.

**Comment by drwpow** [comment]
Thanks for the original question; we’ve gotten this question [in other issues that have been addressed](https://github.com/design-tokens/community-group/issues/240#issuecomment-2294986119), and at this time we’re not open to revisiting the `$` prefix.

TL;DR: this comes from [JSON Schema’s concept of lexical scoping](https://json-schema.org/draft/2020-12/json-schema-core#name-lexical-scope-and-dynamic-s) which uses the `$` keyword to mark reserved words in a dynamic, arbitrarily-nested structure. Another way of looking at it is this specification requires _some_ words to be reserved no matter what; the `$` prefix just gives you more flexibility in naming tokens (including using `value`, `description`, etc.).

**Comment by justinfagnani** [comment]
Sorry if this sounds harsh, but I think that's a very, very unfortunate mis-correlation of the reason for JSON Schema's reason for having a few (8) globally reserved keywords that use a `$` prefix (compared to the majority (50) of their keywords that don't) to your schema here. Those 8 keywords are special because they have to be be able to appear in objects with user-defined keys - but your keywords _don't_.

That in turn is leading to a bigger deign flaw which will make this schema _extremely_ clunky to consume compared to other JSON-based file formats, unless a library layer _specifically undoes_ this choice you've made here. This schema prefixes _all_ keywords instead of creating a more clear structure that doesn't mix keywords and user-defined keys at every level.

As a consumer of this format, I don't want objects with a mix of reserved and auto-chosen names that I have to filter on, I want a stricture that doesn't mix those in the first place.

Consider consuming a structure like this in JavaScript:
```ts
const tokens = {...};
```

Right away I think you have a footgun:
```ts
// Fine for now, because there aren't any top-level keywords _yet_
// Can this break in the future?
const topLevelTokenNames = Object.keys(tokens);
```

Next, let's say you want to get nested tokens of some group:
```ts
const group1 = tokens['Group One'];
const group1Tokens = Object.entries(group1); // oops!
```

Well, that's wrong. What we really need is:
```ts
const group1Tokens = Object.entries(group1).filter(([name]) => !name.startsWith('$'));
```

That's cumbersome, and you have to do it _everywhere_.

It's also cumbersome to tell if an object is even a group. There's no "group" type and because `$types` is allowed on a group, you can use the absence of `$type` to identify a group.

You have to do something like this:
```ts
const isGroup = (token) => Object.keys(token).some((k) => !k.startsWith('$'));
```

It would be much nicer to have an API that naturally separates the schema-defined keys from author-defined keys into cleanly separated collections.

```ts
group1.type; // 'group' so it's easy to find groups vs leafs
group1.itemType; // 'color' - you can still ahve a way to assign the type for all sub-tokens
const group1Tokens = group1.tokens; // no need to filter
```

---

Those examples were in a language that freely lets you mix known an unknown keys like that because of the object type. TypeScript is pretty decent at describing those interfaces, but only because it has a wildly expressive type sysem.

But in many more languages there's a strong difference between object and maps. Any library in such a language is going to have to separate schema-defined and author-defined keys to work at all, undoing this design choice and making working with the library very different from working with the format directly.

---

I really wish you would consider the decisions to mix keys this way. It makes for a harder to write and consume format. And the logic is flawed. There's a reason why almost every JSON format out there uses plain keys, and design tokens _aren't_ some special case that needs special handling.

**Comment by drwpow** [comment]
> Those 8 keywords are special because they have to be be able to appear in objects with user-defined keys - but your keywords don't.

This isn’t correct, or I may be misunderstanding something. The current DTCG format relies on user-defined keys, nested arbitrarily, to infinite depth currently. That is precisely why we have reserved words.

I believe this would only be true _if_ we accepted your proposal to completely rework the schema format to a different structure that was more “AST-ish” and did not allow arbitrary keys. But again, the decision was made long ago to intentionally not follow that format, and at that time we’re not open to changing that.

> That in turn is leading to a bigger deign flaw which will make this schema _extremely_ clunky to consume compared to other JSON-based file formats,

This is opinionated, and I understand this is your experience and I respect it! But I do want to point out that this isn’t a shared opinion, and in my mind isn’t a technical flaw in the spec. Other consumers have had a fairly easy time traversing a schema that follows simple logic:

1. If a node has `$value`, it’s a Token node
2. otherwise, it’s a Group node

The concept of having to keep track of the [“path”](https://tr.designtokens.org/format/#groups) is something that’s also a common requirement of any type of JSON schema, e.g. [JSON pointers](https://opis.io/json-schema/2.x/pointers.html). Again, I respect your experience and don’t disagree that there are tradeoffs with the approach taken here. But I don’t think any are true technical barriers to consuming the specification (or if it is, there are parsers available like [Terrazzo](https://terrazzo.app/)!).
